dot/sync/utils.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2019 ChainSafe Systems (ON) Corp.
+// This file is part of gossamer.
+//
+// The gossamer library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package sync
+
+type reducer func(prevValue interface{}, newValue interface{}) interface{}
+
+// for comp: comp(x, y) = -1: x<y, 0: x==y, 1: x>y
+type comperator func(prevValue interface{}, newValue interface{}) int
+
+// IQR golang outlier detection
+// 	Q25 = 25th_percentile
+// 	Q75 = 75th_percentile
+// 	IQR = Q75 - Q25         // inter-quartile range
+// 	If x >  Q75  + 1.5 * IQR or  x   < Q25 - 1.5 * IQR THEN  x is a mild outlier
+// 	If x >  Q75  + 3.0 * IQR or  x   < Q25 – 3.0 * IQR THEN  x is a extreme outlier
+// Ref: http://www.mathwords.com/o/outlier.htm
+//
+// returns: reducer output
+func RemoveOutlier(sortedArr []interface{}, reducer reducer, compFn comperator, plusFn reducer, minusFn reducer, divideFn reducer, multiplyFn reducer) interface{} {
```

Comment:
nit making it shorter

```suggestion
func RemoveOutlier(sortedArr []interface{}, reducer reducer, compFn comperator, plusFn, minusFn, divideFn, multiplyFn reducer) interface{} {
```